File: csharp/Carballo/Com/Alonsoruibal/Chess/Search/SearchParameters.cs
/*
 * Code converted with Sharpen
 * 
 */using Com.Alonsoruibal.Chess;
using Com.Alonsoruibal.Chess.Log;
using Com.Alonsoruibal.Chess.Search;
using Sharpen;

namespace Com.Alonsoruibal.Chess.Search
{
	public class SearchParameters
	{
		/// <summary>Logger for this class</summary>
		private static readonly Logger logger = Logger.GetLogger("SearchParameters");

		internal int wtime;

		internal int btime;

		internal int winc;

		internal int binc;

		internal int movesToGo;

		internal int depth;

		internal int nodes;

		internal int mate;

		internal int moveTime;

		internal bool infinite;

		internal bool ponder;

		// UCI parameters
		// Remaining time
		// Time increment per move
		// Moves to the next time control
		// Analize x plyes only
		// Search only this number of nodes
		// seatch for mate in mate moves
		// search movetime seconds
		// think infinite
		public virtual bool IsPonder()
		{
			return ponder;
		}

		public virtual void SetPonder(bool ponder)
		{
			this.ponder = ponder;
		}

		public virtual int GetWtime()
		{
			return wtime;
		}

		public virtual void SetWtime(int wtime)
		{
			this.wtime = wtime;
		}

		public virtual int GetBtime()
		{
			return btime;
		}

		public virtual void SetBtime(int btime)
		{
			this.btime = btime;
		}

		public virtual int GetWinc()
		{
			return winc;
		}

		public virtual void SetWinc(int winc)
		{
			this.winc = winc;
		}

		public virtual int GetBinc()
		{
			return binc;
		}

		public virtual void SetBinc(int binc)
		{
			this.binc = binc;
		}

		public virtual int GetMovesToGo()
		{
			return movesToGo;
		}

		public virtual void SetMovesToGo(int movesToGo)
		{
			this.movesToGo = movesToGo;
		}

		public virtual int GetDepth()
		{
			return depth;
		}

		public virtual void SetDepth(int depth)
		{
			this.depth = depth;
		}

		public virtual int GetNodes()
		{
			return nodes;
		}

		public virtual void SetNodes(int nodes)
		{
			this.nodes = nodes;
		}

		public virtual int GetMate()
		{
			return mate;
		}

		public virtual void SetMate(int mate)
		{
			this.mate = mate;
		}

		public virtual int GetMoveTime()
		{
			return moveTime;
		}

		public virtual void SetMoveTime(int moveTime)
		{
			this.moveTime = moveTime;
		}

		public virtual bool IsInfinite()
		{
			return infinite;
		}

		public virtual void SetInfinite(bool infinite)
		{
			this.infinite = infinite;
		}

		/// <summary>TODO elaborate a bit</summary>
		/// <returns></returns>
		public virtual long CalculateMoveTime(Board board)
		{
			if (infinite)
			{
				return 999999999;
			}
			if (moveTime != 0)
			{
				return moveTime;
			}
			int calctime = 0;
			if (board.GetTurn())
			{
				if (wtime > 0)
				{
					calctime = wtime / 40 + winc;
				}
			}
			else
			{
				if (btime > 0)
				{
					calctime = btime / 40 + binc;
				}
			}
			logger.Debug("Thinking for " + calctime + "Ms");
			return calctime;
		}

		public static SearchParameters Get(int moveTime)
		{
			SearchParameters searchParameters = new SearchParameters();
			searchParameters.SetMoveTime(moveTime);
			return searchParameters;
		}
	}
}

File: core/src/main/java/com/alonsoruibal/chess/search/SearchParameters.java
package com.alonsoruibal.chess.search;

import com.alonsoruibal.chess.log.Logger;

public class SearchParameters {
	/**
	 * Logger for this class
	 */
	private static final Logger logger = Logger.getLogger("SearchParameters");

	// UCI parameters
	// Remaining time
	int wtime, btime;
	// Time increment per move
	int winc, binc;
	// Moves to the next time control
	int movesToGo;
	// Analize x plyes only
	int depth = Integer.MAX_VALUE;
	// Search only this number of nodes
	int nodes = Integer.MAX_VALUE;
	// Search for mate in mate moves
	int mate;
	// Search movetime milliseconds
	int moveTime = Integer.MAX_VALUE;
	// Think infinite
	boolean infinite;
	boolean ponder;

	boolean manageTime;

	public boolean isPonder() {
		return ponder;
	}

	public void setPonder(boolean ponder) {
		this.ponder = ponder;
	}

	public int getWtime() {
		return wtime;
	}

	public void setWtime(int wtime) {
		this.wtime = wtime;
	}

	public int getBtime() {
		return btime;
	}

	public void setBtime(int btime) {
		this.btime = btime;
	}

	public int getWinc() {
		return winc;
	}

	public void setWinc(int winc) {
		this.winc = winc;
	}

	public int getBinc() {
		return binc;
	}

	public void setBinc(int binc) {
		this.binc = binc;
	}

	public int getMovesToGo() {
		return movesToGo;
	}

	public void setMovesToGo(int movesToGo) {
		this.movesToGo = movesToGo;
	}

	public int getDepth() {
		return depth;
	}

	public void setDepth(int depth) {
		this.depth = depth;
	}

	public int getNodes() {
		return nodes;
	}

	public void setNodes(int nodes) {
		this.nodes = nodes;
	}

	public int getMate() {
		return mate;
	}

	public void setMate(int mate) {
		this.mate = mate;
	}

	public int getMoveTime() {
		return moveTime;
	}

	public void setMoveTime(int moveTime) {
		this.moveTime = moveTime;
	}

	public boolean isInfinite() {
		return infinite;
	}

	public void setInfinite(boolean infinite) {
		this.infinite = infinite;
	}

	/**
	 * Used to detect if it can add more time in case of panic or apply other heuristics to reduce time
	 *
	 * @return true if the engine is responsible of managing the remaining time
	 */
	public boolean manageTime() {
		return manageTime;
	}

	/**
	 * Time management routine
	 * @param panicTime is set to true when the score fails low in the root node by 100
	 *
	 * @return the time to think, or Long.MAX_VALUE if it can think an infinite time
	 */
	public long calculateMoveTime(boolean engineIsWhite, long startTime, boolean panicTime) {
		manageTime = false;
		if (ponder || infinite || depth < Integer.MAX_VALUE || nodes < Integer.MAX_VALUE) {
			return Long.MAX_VALUE;
		}
		if (moveTime != Integer.MAX_VALUE) {
			return startTime + moveTime;
		}
		manageTime = true;

		int calcTime = 0;
		int timeAvailable = engineIsWhite ? wtime : btime;
		int timeInc = engineIsWhite ? winc : binc;
		if (timeAvailable > 0) {
			calcTime = timeAvailable / 25;
		}
		if (panicTime) { // x 4
			calcTime = calcTime << 2;
		}
		calcTime = Math.min(calcTime, timeAvailable >>> 3); // Never consume more than time / 8
		calcTime += timeInc;

		logger.debug("Thinking for " + calcTime + "Ms");
		return startTime + calcTime;
	}

	public static SearchParameters get(int moveTime) {
		SearchParameters searchParameters = new SearchParameters();
		searchParameters.setMoveTime(moveTime);
		return searchParameters;
	}
}
